lib/spack/spack/build_systems/autotools.py

```diff
@@ -140,16 +158,46 @@ def patch(self):
 
     def autoreconf(self, spec, prefix):
         """Not needed usually, configure should be already there"""
-        pass
+        # If configure exists nothing needs to be done
+        if os.path.exists(self.configure_abs_path):
+            return
+        # Else try to regenerate it
+        autotools = ['m4', 'autoconf', 'automake', 'libtool']
+        missing = [x for x in autotools if x not in spec]
+        if missing:
+            msg = 'Cannot generate configure: missing dependencies {0}'
+            raise RuntimeError(msg.format(missing))
+        tty.msg('Configure script not found: trying to generate it')
+        tty.warn('*********************************************************')
+        tty.warn('* If the default procedure fails, consider implementing *')
+        tty.warn('*        a custom AUTORECONF phase in the package       *')
+        tty.warn('*********************************************************')
+        with working_dir(self.configure_directory):
+            m = inspect.getmodule(self)
+            m.libtoolize()
+            m.aclocal()
+            m.autoconf()
+            autogen = Executable('./autogen.sh')
+            autogen()
```

Comment:
According to docs:
- `aclocal` [adds m4 macros](https://www.gnu.org/software/automake/manual/html_node/aclocal-Invocation.html) that may be needed by `configure.ac` in the folder aclocal.m4
- `libtoolize` [provides a way to add support for libtool](https://www.gnu.org/software/libtool/manual/html_node/Invoking-libtoolize.html)
- `autoconf` [creates a new configure](https://www.gnu.org/software/autoconf/manual/autoconf-2.67/html_node/autoconf-Invocation.html) from `aclocal.m4` and `configure.ac`
- `autoreconf` [does a few of the above things](https://www.gnu.org/software/autoconf/manual/autoconf-2.68/html_node/autoreconf-Invocation.html) repeatedly to generate a configure (but I remember I saw cases were the order was not right and plain autoreconf was failing)
- to my surprise `autogen.sh` seems not to be a generated file, but an [hand maintained one](https://www.sourceware.org/autobook/autobook/autobook_43.html), even though [it is mentioned in the FAQ](https://www.gnu.org/software/automake/faq/autotools-faq.html#What-does-_002e_002fbootstrap-or-_002e_002fautogen_002esh-do_003f).

Given all of the above I would remove:
```python
autogen()
```
and substitute it with:
```python
autoreconf('--install', '--verbose', '--force')
```
and hope that @mathstuf and colleagues will take over the market with CMake. Does it sound good @adamjstewart ?